src/Storages/MergeTree/IntersectionsIndexes.h

```diff
@@ -0,0 +1,206 @@
+#pragma once
+
+#include <fmt/format.h>
+#include <Storages/MergeTree/RequestResponse.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+/// A boundary of a segment (left or right)
+struct PartToRead
+{
+    PartBlockRange range;
+    String name;
+
+    bool operator==(const PartToRead & rhs) const
+    {
+        return range == rhs.range && name == rhs.name;
+    }
+
+    bool operator<(const PartToRead & rhs) const
+    {
+        /// We allow only consecutive non-intersecting ranges
+        if (rhs.range.begin > range.begin && rhs.range.begin < range.end)
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Bad ranges");
+        return range.begin < rhs.range.begin && range.end <= rhs.range.begin;
+    }
+};
+
+
+struct PartSegments
+{
+    enum class IntersectionResult
+    {
+        NO_INTERSECTION,
+        EXACTLY_ONE_INTERSECTION,
+        REJECT
+    };
+
+    void addPart(PartToRead part) { segments.insert(std::move(part)); }
+
+    IntersectionResult getIntersectionResult(PartToRead part)
+    {
+        bool intersected_before = false;
+        for (const auto & segment: segments)
+        {
+            auto are_intersect = [](auto & x, auto & y)
+            {
+                /// <= is important here, because we are working with segments [a, b]
+                if ((x.begin <= y.begin) && (y.begin <= x.end))
+                    return true;
+                if ((y.begin <= x.begin) && (x.begin <= y.end))
+                    return true;
+                return false;
+            };
+
+            if (are_intersect(segment.range, part.range))
+            {
+                /// We have two or possibly more intersections
+                if (intersected_before)
+                    return IntersectionResult::REJECT;
+
+                /// We have intersection with part with different name
+                /// It could happens if we have merged part on one replica
+                /// but not on another.
+                if (segment.name != part.name)
+                    return IntersectionResult::REJECT;
+
+                intersected_before = true;
+            }
+        }
+
+        return intersected_before ? IntersectionResult::EXACTLY_ONE_INTERSECTION : IntersectionResult::NO_INTERSECTION;
```

Comment:
And why do we allow a range like `[1 3]` if there is a range like `[1 5]`?

From my point of view, we should either:
* find an exact match and return it
* find that range does not intersect with current ranges, and insert it
* find any intersection and reject

Reply:
Yes, that is a bug. We have to allow an intersection with exact match, otherwise reject it or tell that there are no intersections.